src/support/small_set.h

```diff
@@ -0,0 +1,298 @@
+          flexible.insert(fixed[i]);
+        }
+        flexible.insert(x);
+        assert(!usingFixed());
+        usedFixed = 0;
+      }
+    } else {
+      flexible.insert(x);
+    }
+  }
+
+  void erase(const T& x) {
+    if (usingFixed()) {
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          // We found the item; erase it by moving the final item to replace it
+          // and truncating the size.
+          usedFixed--;
+          fixed[i] = fixed[usedFixed];
+          return;
+        }
+      }
+    } else {
+      flexible.erase(x);
+    }
+  }
+
+  size_t count(const T& x) {
+    if (usingFixed()) {
+      // Do a linear search.
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          return 1;
+        }
+      }
+      return 0;
+    } else {
+      return flexible.count(x);
+    }
+  }
+
+  size_t size() const {
+    if (usingFixed()) {
+      return usedFixed;
+    } else {
+      return flexible.size();
+    }
+  }
+
+  bool empty() const { return size() == 0; }
+
+  void clear() {
+    usedFixed = 0;
+    flexible.clear();
+  }
+
+  bool operator==(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    if (usingFixed()) {
+      if (usedFixed != other.usedFixed) {
+        return false;
+      }
+      for (size_t i = 0; i < usedFixed; i++) {
+        bool found = false;
+        for (size_t j = 0; j < usedFixed; j++) {
+          if (fixed[i] == other.fixed[j]) {
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return flexible == other.flexible;
+  }
+
+  bool operator!=(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    return !(*this == other);
+  }
+
+  // iteration
+
+  template<typename Parent, typename Iterator, typename FlexibleIterator>
+  struct IteratorBase : public std::iterator<std::forward_iterator_tag, T> {
+    typedef T value_type;
+    typedef long difference_type;
+    typedef T& reference;
+
+    Parent* parent;
+
+    // Whether we are using fixed storage in the parent. When doing so we have
+    // the index in fixedIndex. Otherwise, we are using flexible storage, and we
+    // will use flexibleIterator.
+    bool usingFixed;
+    size_t fixedIndex;
+    FlexibleIterator flexibleIterator;
+
+    IteratorBase(Parent* parent)
+      : parent(parent), usingFixed(parent->usingFixed()) {}
+
+    void setBegin() {
+      if (usingFixed) {
+        fixedIndex = 0;
+      } else {
+        flexibleIterator = parent->flexible.begin();
+      }
+    }
+
+    void setEnd() {
+      if (usingFixed) {
+        fixedIndex = parent->size();
+      } else {
+        flexibleIterator = parent->flexible.end();
+      }
+    }
+
+    bool operator==(const Iterator& other) const {
+      if (parent != other.parent) {
+        return false;
+      }
+      assert(usingFixed == other.usingFixed);
```

Comment:
I guess it would be UB if the mode were different (it would mean we added or removed from the set meanwhile), so maybe we don't need to check this? But it seems simple to assert on it and then the code below is obviously correct I think.

Reply:
I don't follow... Can't users of `SmallSet` or `SmallSet::iterator` freely compare whatever iterators they have? If `usingFixed != other.usingFixed`, shouldn't we just return `false`? Not sure what you mean about UB..

Reply:
Sorry, I meant: since we've already found that the parent is identical, then these are iterators to the same set. In that case their mode must be identical, unless one was created, then the set modified (enough to change the mode), then another created. But thinking on it more, I guess that isn't always UB?

Reply:
Calling `insert` on std::set does not invalidate iterators, so arguably it should not invalidate iterators here, either. But there's no good way to keep an iterator valid across the transition from `fixed` to `flexible`, so it is probably cleaner to just say that all modifications invalidate iterators and keep this assert.

Reply:
I'm worried about even trying to handle that case. How about just erroring on it - more than an assertion, even? Then there is no risk of things going silently wrong.